AUT-313: break up too-large-for-GCP random reads before sending them to libkmsp11

bhearsum opened the pull request:
This allows for the reading of random bytes of any size, which is needed by various operations that Autograph performs.

jmhodges commented on signer/hsm.go:
```diff
@@ -107,6 +107,34 @@ type GCPHSM struct {
+	var result []byte
+	need := len(data)
+	for need > 0 {
+		next := make([]byte, 1024)
```
`min(need, 1024)` or similar?

bhearsum replied:
Yeah, this needs some more iteration (and tests)!

bhearsum commented:
@say-yawn prompted me to add some test cases for failures; I'm moving back to draft while I do that.

alexcottner commented on signer/hsm.go:
```diff
@@ -107,6 +107,36 @@ type GCPHSM struct {
+		bytesRead, err = ubrr.rr.Read(next)
+		n += bytesRead
+		if err != nil {
+			return
```
Is there any way to not lose the error here and return/log it?

bhearsum replied:
AIUI, we're returning that error (the `err` variable is bound to the return as a "named result parameter": https://go.dev/doc/effective_go#named-results)

I think this is verified by the new test where we mock in a specific error that `GenerateRandom` returns, and verify that it comes out of UnlimitedBytesRandReader.Read: https://github.com/mozilla-services/autograph/pull/1032/files#diff-a43f626c786b027528255923aee643b55513fb5d596a69e082c63dba5b9ccfb9R833-R857

alexcottner replied:
Ah, overlooked that thank you.

bhearsum replied:
Sure. I thought they were idiomatic, but I have no real preference.

jmhodges replied:
Some folks in Go-land like them, but I think the community is mostly on the side of "too easy to have shadowing problems".

jmhodges commented:
Good patch! I had some small things that I could get push back on.

jmhodges commented on signer/hsm.go:
```diff
@@ -107,6 +107,36 @@ type GCPHSM struct {
+		result = append(result, next...)
+		need -= bytesRead
+	}
+	copy(data, result)
```
Enh, this isn't worth it, probably.

bhearsum commented on signer/signer_test.go:
```diff
@@ -662,3 +663,208 @@ func TestMakeKeyGCPRSA(t *testing.T) {
 	}
 }
+
+var UNLIMITEDBYTESREADERTESTCASES = []struct {
```
I can adjust this; I'm going to change the style for the whole file though to maintain consistency.

bhearsum commented on signer/signer_test.go:
```diff
@@ -662,3 +663,208 @@ func TestMakeKeyGCPRSA(t *testing.T) {
+	defer crypto11.Close()
+
+	for _, testcase := range UNLIMITEDBYTESREADERTESTCASES {
+		for _, size := range testcase.expectedPkcs11RandReaderBytes {
```
I could certainly add this, but unless I'm missing something it would require me to have the mock (or an additional mock) behave differently than the real method, which doesn't seem ideal.

A different thing we could do is to have `unlimitedBytesRandReader.Read` mirror the behaviour of the underlying method, and have `n` always be 0 or `len(data)`. I'm not super excited to do that either, but the more I think about it, the more it actually seems like the most correct thing to do?

jmhodges replied:
If it returns 0, it'd return an error, right? You could have the mock return an error.

Just an FYI, there is an [InOrder](https://pkg.go.dev/go.uber.org/mock/gomock#InOrder) api that makes it a lil easier to interleave responses.

bhearsum replied:
That's my understanding - yes.

Oh wait, I think I just realized you were asking to make sure that our code behaves correctly -- not that it returns > 0 on error. Yes of course, that's totally doable (and sensible).

bhearsum replied:
Oh, this is already covered though? The mock is on `GenerateRandom`, which AFAIK might return a non-zero number of bytes, but also an error. https://github.com/mozilla-services/autograph/pull/1032/files#diff-a43f626c786b027528255923aee643b55513fb5d596a69e082c63dba5b9ccfb9R840-R845 covers this, I believe?

jmhodges replied:
SGTM!

jmhodges commented on signer/hsm.go:
```diff
@@ -107,6 +107,36 @@ type GCPHSM struct {
+		if err != nil {
+			return n, err
+		}
+		data = append(data, next...)
```
I'm not sure how this is working correctly. It only appends to data, but does not overwrite existing data in `data`. That doesn't seem correct?

jmhodges replied:
I think going back to the copying version is fine, but we should figure out why the tests were able to pass.

bhearsum commented:
@jmhodges - we've had a bunch of back and forth here. I _think_ this is done now, so I'll send it back. Feel free to catch me on Slack if we need to hash anything out still.

jmhodges commented:
> @jmhodges - we've had a bunch of back and forth here. I _think_ this is done now, so I'll send it back. Feel free to catch me on Slack if we need to hash anything out still.

I don't see any changes in the testing which makes me nervous that our testing is incomplete. What am I missing?

bhearsum commented on signer/signer_test.go:
```diff
@@ -662,3 +664,262 @@ func TestMakeKeyGCPRSA(t *testing.T) {
+				bytes.Repeat([]byte("B"), 1024)...,
+			),
+			bytes.Repeat([]byte{0}, 2952)...,
+		),
```
There may be a better way to generate these inline, but I couldn't find it.

jmhodges replied:
yeah, the only think i could think of was a "pre-processing" for-loop that turned sizes into slices and did byte math with the iterator (`byte('a')+i`) between the test cases and the actual uses. But that has the downsize of not being explicit in what folks should see. I like this better!

bhearsum commented:
@jmhodges Sorry about all the confusion around the tests. I updated them to test actual bytes returned rather than just the lengths. This also, of course, uncovered the fact that we weren't actually returning partially satisfied requests in UnlimitedBytesRandReader.Read, so I fixed that (and got rid of the unnecessary `copy` on the way).
